hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/ReadHadoopNewApiP.java

```diff
@@ -0,0 +1,284 @@
+import static com.hazelcast.jet.impl.util.Util.uncheckCall;
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * See {@link HadoopSources#inputFormat}.
+ */
+public final class ReadHadoopNewApiP<K, V, R> extends AbstractProcessor {
+
+    private static final Class<?>[] EMPTY_ARRAY = new Class[0];
+
+    private final Configuration configuration;
+    private final InputFormat inputFormat;
+    private final Traverser<R> trav;
+    private final BiFunctionEx<K, V, R> projectionFn;
+    private final BetterByteArrayOutputStream cloneBuffer = new BetterByteArrayOutputStream();
+    private final DataOutputStream cloneBufferDataOutput = new DataOutputStream(cloneBuffer);
+
+    private InternalSerializationService serializationService;
+    private RecordReader<K, V> reader;
+
+    private ReadHadoopNewApiP(
+            @Nonnull Configuration configuration,
+            @Nonnull InputFormat inputFormat,
+            @Nonnull List<InputSplit> splits,
+            @Nonnull BiFunctionEx<K, V, R> projectionFn
+    ) {
+        this.configuration = configuration;
+        this.inputFormat = inputFormat;
+        this.trav = traverseIterable(splits)
+                .flatMap(this::traverseSplit);
+        this.projectionFn = projectionFn;
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        HazelcastInstanceImpl instance = (HazelcastInstanceImpl) context.jetInstance().getHazelcastInstance();
+        serializationService = instance.getSerializationService();
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return false;
+    }
+
+    @Override
+    public boolean complete() {
+        return emitFromTraverser(trav);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (reader != null) {
+            reader.close();
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private Traverser<R> traverseSplit(InputSplit split) {
+        try {
+            TaskAttemptContextImpl attemptContext = new TaskAttemptContextImpl(configuration, new TaskAttemptID());
+            reader = inputFormat.createRecordReader(split, attemptContext);
+            reader.initialize(split, attemptContext);
+        } catch (IOException | InterruptedException e) {
+            throw sneakyThrow(e);
+        }
+
+        return () -> {
+            try {
+                while (reader.nextKeyValue()) {
+                    // we clone the key/value if configured so because some of the
+                    // record-readers return the same object for `reader.getCurrentKey()`
+                    // and `reader.getCurrentValue()` which is mutated for each `reader.nextKeyValue()`
+                    R projectedRecord = projectionFn.apply(reader.getCurrentKey(), reader.getCurrentValue());
+                    if (projectedRecord != null) {
+                        return copyIfConfigured(projectedRecord);
```

Comment:
also, this method is called for each record, which will do a hash lookup in the configuration